core/src/main/scala/com/banno/kafka/consumer/ConsumerOps.scala

```diff
@@ -353,4 +353,101 @@ case class ConsumerOps[F[_], K, V](consumer: ConsumerApi[F, K, V]) {
       .recordStream(pollTimeout)
       .evalMap(r => process(r) <* consumer.commitSync(r.nextOffset))
 
+  /** Returns a stream that processes records using the specified function,
+    * committing offsets for successfully processed records, either after
+    * processing the specified number of records, or after the specified time
+    * has elapsed since the last offset commit. If the processing function
+    * returns a failure, offsets for records successfully processed before that
+    * failure will be committed, and then the stream will halt with that
+    * failure. This is at-least-once processing: after a restart, the record
+    * that failed will be reprocessed. In some use cases this pattern is more
+    * appropriate than just using auto-offset-commits, since it will not commit
+    * offsets for failed records when the consumer is closed. The consumer must
+    * be configured to disable offset auto-commits.
+    */
+  def processingAndCommitting[A](
+      pollTimeout: FiniteDuration,
+      maxRecordCount: Long = 1000L,
+      maxElapsedTime: FiniteDuration = 60.seconds,
+  )(
+      process: ConsumerRecord[K, V] => F[A]
+  )(implicit C: Clock[F], S: Concurrent[F]): Stream[F, A] =
+    for {
+      state <- Stream.eval(
+        C.realTime
+          .flatMap(now =>
+            Ref.of[F, OffsetCommitState](OffsetCommitState.empty(now))
+          )
+      )
+      record <- consumer.recordStream(pollTimeout)
+      a <- Stream.eval(
+        process(record)
+          .onError(_ =>
+            // we can still commit offsets that were successfully processed, before this stream halts, consumer is closed, etc
+            // this is still at-least-once processing, but minimizes the amount of reprocessing after restart
+            state.get.flatMap(s => consumer.commitSync(s.nextOffsets))
+          )
+      )
+      s <- Stream.eval(state.updateAndGet(_.update(record)))
+      now <- Stream.eval(C.realTime)
+      () <- Stream.eval(
+        s.needToCommit(maxRecordCount, now, maxElapsedTime)
+          .traverse_(os =>
+            consumer.commitSync(os) *>
+            state.update(_.reset(now))
+          )
+      )
+    } yield a
+
+  case class OffsetCommitState(
+      offsets: Map[TopicPartition, Long],
+      recordCount: Long,
+      lastCommitTime: FiniteDuration,
+  ) {
+    def update(record: ConsumerRecord[_, _]): OffsetCommitState =
+      copy(
+        offsets = offsets + (new TopicPartition(
+          record.topic,
+          record.partition,
+        ) -> record.offset),
+        recordCount = recordCount + 1,
+      )
+
+    def needToCommit(
+        maxRecordCount: Long,
+        now: FiniteDuration,
+        maxElapsedTime: FiniteDuration,
+    ): Option[Map[TopicPartition, OffsetAndMetadata]] = {
+      if (
```

Comment:
Not sure if it reads better or not, but another way to write this would be
```scala
(
  recordCount >= maxRecordCount || (now - lastCommitTime) >= maxElapsedTime
).guard[Option]
  .as(nextOffsets)
```

Reply:
Nice! I do like the way that reads better, and it formats better too. Refactored in https://github.com/Banno/kafka4s/pull/782/commits/4ea9f79a5af6d66d37dcb5efe01cb9c39447609d.